runtime/src/main/java/org/corfudb/runtime/view/replication/ChainReplicationProtocol.java

```diff
@@ -0,0 +1,155 @@
+     *
+     */
+    protected void recover(long globalAddress) {
+        // In chain replication, we started writing from the head,
+        // and propagated down to the tail. To recover, we start
+        // reading from the head, which should have the data
+        // we are trying to recover
+        int numUnits = layout.getSegmentLength(globalAddress);
+        log.debug("Recover[{}]: read chain head {}/{}", globalAddress, 1, numUnits);
+        ILogData ld = CFUtils.getUninterruptibly(layout.getLogUnitClient(globalAddress, 0)
+                .read(globalAddress)).getReadSet().getOrDefault(globalAddress, null);
+        // If nothing was at the head, this is a bug and we
+        // should fail with a runtime exception, as there
+        // was nothing to recover - if the head was removed
+        // due to a reconfiguration, a network exception
+        // would have been thrown and the client should have
+        // retried it's operation (in this case of a write,
+        // it should have read to determine whether the
+        // write was successful or not.
+        if (ld == null) {
+            throw new RuntimeException();
+        }
+        // now we go down the chain and write, ignoring any overwrite exception we get.
+        for (int i = 1; i < numUnits; i++) {
+            log.debug("Recover[{}]: write chain {}/{}", layout, i + 1, numUnits);
+            // In chain replication, we write synchronously to every unit
+            // in the chain.
+            try {
+                CFUtils.getUninterruptibly(
+                        layout.getLogUnitClient(globalAddress, i)
+                                // TODO: Currently, this call won't work until the log unit client is refactored.
+                                .write(globalAddress, null, null, ld, null)
+                        , OverwriteException.class);
+                // We successfully recovered a write to this member of the chain
+                log.debug("Recover[{}]: recovered write at chain {}/{}", layout, i + 1, numUnits);
+            } catch (OverwriteException oe) {
+                // This member already had this data (in some cases, the write might have
+                // been committed to all members, so this is normal).
+                log.debug("Recover[{}]: overwritten at chain {}/{}", layout, i + 1, numUnits);
+            }
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    protected void holeFill(long globalAddress) {
+        int numUnits = layout.getSegmentLength(globalAddress);
+        for (int i = 0; i < numUnits; i++) {
+            log.trace("fillHole[{}]: chain {}/{}", globalAddress, i + 1, numUnits);
+            // In chain replication, we write synchronously to every unit in
+            // the chain.
+            try {
+                CFUtils.getUninterruptibly(layout.getLogUnitClient(globalAddress, i)
+                        .fillHole(globalAddress), OverwriteException.class);
```

Comment:
Not ok. This is precisely the problem in the code right now. Overwrite occurs at the head,  but no write has propogated to the tail. The client must recover if it gets an overwrite exception from any log unit during a write to adopt the other value.